py_zipkin/zipkin.py

```diff
@@ -709,3 +713,125 @@ def create_http_headers_for_new_span(context_stack=None, tracer=None):
+    if not is_valid_trace_id(bits[0]) or not is_valid_span_id(bits[1]):
+        return None
+    if len(bits) == 4:
+        if not is_valid_span_id(bits[3]):
+            return None
+        parent_span_id = bits[3]
+    else:
+        parent_span_id = None
+    if len(bits) >= 3 and bits[2]:
+        # We have 1==Accept, 0==Deny, d==Debug
+        if bits[2] == 'd' or bits[2] == '1' or bits[2] == 'true':
+            is_sampled = True
+        elif bits[2] == '0' or bits[2] == 'false':
+            is_sampled = False
+        else:
+            # Not valid according to spec
+            return None
+    else:
+        # sample flag missing; means "Defer" and we're responsible for
+        # rolling fresh dice
+        is_sampled = choose_sampling(sample_rate)
+    return ZipkinAttrs(bits[0], bits[1], parent_span_id,
+                       '0',  # Flags aren't a thing in single-header
+                       is_sampled)
+
+
+def extract_zipkin_attrs_from_headers(headers, sample_rate=100.0,
+                                      use_128bit_trace_id=False):
+    """
+    Implements extraction of B3 headers per:
+        https://github.com/openzipkin/b3-propagation
+
+    The input headers can be any dict-like container that supports "in"
+    membership test and a .get() method that accepts a default value.
+
+    Returns a ZipkinAttrs instance or None
+    """
+    if 'b3' in headers:
+        return _extract_zipkin_attrs_from_single_header(
+            headers['b3'], sample_rate, use_128bit_trace_id)
+
+    # No single header, let's get old-school
+    trace_id = headers.get('X-B3-TraceId', None)
+    span_id = headers.get('X-B3-SpanId', None)
+    parent_span_id = headers.get('X-B3-ParentSpanId', None)
+    sampled_str = headers.get('X-B3-Sampled', None)
+    flags_str = headers.get('X-B3-Flags', '0')
+    is_debug = flags_str == '1'
+    sample_decision_only = (
+        all(h is None for h in (trace_id, span_id, parent_span_id)) and
+        (sampled_str or is_debug))
+
+    if not sample_decision_only:
+        if not is_valid_trace_id(trace_id):
+            return None
+        if not is_valid_span_id(span_id):
+            return None
+        if parent_span_id and not is_valid_span_id(parent_span_id):
+            return None
+
+    if is_debug:
+        if sample_decision_only:
+            return create_attrs_for_span(
+                sample_rate=100.0,
+                use_128bit_trace_id=use_128bit_trace_id)
+        is_sampled = True
+    elif sampled_str:
+        # We have 1==Accept, 0==Deny
+        # Also, "While you shouldn't encode X-B3-Sampled as true or false, a
+        #        lenient implementation may accept them."
+        if sampled_str == '1' or sampled_str == 'true':
+            is_sampled = True
+        elif sampled_str == '0' or sampled_str == 'false':
+            is_sampled = False
+        else:
+            # Not valid according to spec
+            return None
```

Comment:
Maybe log a warning in these cases where the headers are not valid?

Reply:
Done (throughout).  I picked `log.warning(...)` let me know if another level is more appropriate.